soroban-sdk/src/unwrap.rs

```diff
@@ -32,3 +34,19 @@ impl<T, E: core::fmt::Debug> UnwrapOptimized for Result<T, E> {
         self.unwrap()
     }
 }
+
+pub trait UnwrapInfallible {
+    type Output;
+    fn unwrap_infallible(self) -> Self::Output;
+}
+
+impl<T> UnwrapInfallible for Result<T, Infallible> {
+    type Output = T;
+
+    fn unwrap_infallible(self) -> Self::Output {
+        match self {
+            Ok(ok) => ok,
+            Err(never) => match never {},
```

Comment:
I don’t get what’s happening here. What gets returned in the Err case? It looks like `Self::Output` is the type that's expected in the `Ok` case.

Reply:
Yeah, this is a subtlety of the way empty types work. Which might require backing up even a further step to discuss what an empty type _is_ and what it's for. Empty types are types that have no inhabitants, so they're like "one less inhabitant than the `()` type". Whereas with `()` you can only construct one possible value (so it takes zero space, every `()` is identical) with an empty type you can _never_ construct a value of the type. It's what the type system assigns to definitely-diverging calculations: infinite loops, functions that exit the program, etc. Evaluation-that-does-not-return, impossible control paths. Or in our case "what happens when you call a host function from the VM guest and the host has an error". You get a `Result<T, Infallible>` where the `Err` case is defined as "can't happen" / can't be observed, because if there was an error in the host the host would trap the guest's VM, preventing it from observing anything further.

Rust has two "empty types", one called `!` which is built into the language, and one called `Infallible` which is defined in the core library as an empty enum for just our use case: `enum Infallible {}`. The latter only exists due to weird little technical limitations in the way the former was originally implemented -- you can't put `!` in a data structure currently -- as far as I can tell; the plan is to merge them when they clean out all the cruft in the implementation. So you can/should just read the two as "mostly synonymous" and our result type here as mostly meaning `Result<T, !>`, except that a few cases you have to do weird conversions between the two. In this case we are doing a weird conversion between `Infallible` and `!` by matching on the `Infallible` in a match expression that has no arms and can never succeed (or even commence); this (as far as I can tell, I just copied this from elsewhere) converts it from a "datatype that isn't given enough special treatment" to a control-flow idiom that rust knows to have type `!`.

Why do this here? Because we need a piece of the special treatment `!` gets. One of the quirks of empty types in general (languages vary, but this isn't too uncommon) is that they _unify with everything_. If you have an expression of type `!` it is allowed to be used as any other type `T`, because in the process of evaluating the `!` you're saying control will never actually complete, so nobody would ever be able to determine the absence of a `T` value. Essentially the type system is willing to let you use any possible never-returning as a promise that when you do return you'll bring some `T` with you. It's weird but it's how it works and it's what's happening here.

I'll push a comment that explains this in more detail now. I want it to be clear to readers, and I agree it's a bit mysterious as-is.

Reply:
Thanks for the in-depth explanation!

Reply:
@graydon Thank you for the explanation. I understand the syntactical part of receiving an `Infallible` and needing to convert it to `!` so that it can be treated as any `T` by the type system. 
But I'm still puzzled about what actually happens next. If I may use a simple code example in sdk's `vec.rs`:

```
    pub fn is_empty(&self) -> bool {
        let env = self.env();
        let val = env.vec_len(self.obj).unwrap_infallible();
        val.is_u32_zero()
    }
```

If `env.vec_len` returns an `Err(Infallible)`, which `unwrap_infallible()` converts to an `!` (or is it `u32` as the type system actually sees as?). What happens next? Does the code immediate panic/crash because an `!` has been encountered? Or does the next line execute which calls `is_u32_zero` on `!` which then triggers the system to panic/crash?